src/action/install_date.rs

```diff
@@ -0,0 +1,284 @@
+    // on installation:
+    // * /var/log/installer/: The location of installation logs for Debian-based
+    //   distributions.
+    // * /root/install.log: The location of installation logs for RHEL-based
+    //   distributions.
+    // * /etc/hostname: In most cases, we may assume that the hostname didn't
+    //   change after installation.
+    // * /lost+found: This method was used by Python version on GRR client, so
+    //   leaving it here.
+    const CANDIDATES: [&str; 5] = [
+        "/var/log/installer/syslog",
+        "/var/log/installer/status",
+        "/root/install.log",
+        "/etc/hostname",
+        "/lost+found",
+    ];
+    if let Some(time) = get_modified_time(CANDIDATES.iter()) {
+        return Some(time);
+    }
+
+    // We tried our best and all the methods have failed, so just give up.
+    None
+}
+
+/// Obtains system install date (OS X version).
+///
+/// This function searches for some files that were probably unchanged since
+/// the installation.
+///
+/// This function returns `None` in case of errors.
+#[cfg(target_os = "macos")]
+fn get_install_time() -> Option<SystemTime> {
+    // Here, we use the same way as Python version of GRR client does. We just
+    // check the modification time for some of the paths
+    const CANDIDATES: [&str; 3] = [
+        "/var/log/CDIS.custom",
+        "/var",
+        "/private",
+    ];
+    get_modified_time(CANDIDATES.iter())
+}
+
+/// Obtains system install date (Windows version).
+///
+/// To obtain install date, this function searches the corresponding value in
+/// Windows registry.
+///
+/// This function returns `None` in case of errors.
+#[cfg(target_os = "windows")]
+fn get_install_time() -> Option<SystemTime> {
+    use winreg::RegKey;
+
+    // Don't use winreg::enums::KEY_WOW64_64KEY since it breaks on Windows 2000
+    let hklm = RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+    let install_time = hklm
+        .open_subkey("Software\\Microsoft\\Windows NT\\CurrentVersion").ok()?
+        .get_value::<u32, _>("InstallDate").ok()?;
+    Some(SystemTime::UNIX_EPOCH + Duration::from_secs(install_time.into()))
+}
+
+/// Handles requests for the install date action.
+pub fn handle<S: Session>(session: &mut S, _: ()) -> session::Result<()> {
+    session.reply(Response {time: get_install_time()})?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use humantime::Timestamp;
+
+    #[test]
+    fn test_install_date() {
+        let mut session = session::test::Fake::new();
+        assert!(handle(&mut session, ()).is_ok());
+        assert_eq!(session.reply_count(), 1);
+        let response: &Response = session.reply(0);
+        let time = response.time.unwrap();
+
+        // We assume here that the tests won't be run for very old systems
+        // installed before 01.01.2000.
+        let lower_limit: Timestamp = "2000-01-01T00:00:00Z".parse().unwrap();
```

Comment:
I am not sure if setting some arbitrary lower limit other than epoch makes sense (because it's hard to decide what should it be exactly).

Reply:
The dates that are much higher than epoch (like 2000-01-01) can potentially catch issues with wrong measurement units in the code, so I would prefer leaving it as is.

Reply:
These issues should be covered by the upper-limit check, no? Of course not all of them, but then you could argue that some issues are not caught be the 2000.01.01 either. And the choice of it seems rather arbitrary.

Adding such assumptions, while can be seen as quite innocent here, sets a precedent for more serious cases.